crates/ruff_python_ast/src/nodes.rs

```diff
@@ -3196,37 +3254,40 @@ pub struct Parameters {
 }
 
 impl Parameters {
-    /// Returns the [`ParameterWithDefault`] with the given name, or `None` if no such [`ParameterWithDefault`] exists.
-    pub fn find(&self, name: &str) -> Option<&ParameterWithDefault> {
+    /// Returns an iterator over all non-variadic parameters included in this [`Parameters`] node.
+    ///
+    /// The variadic parameters (`.vararg` and `.kwarg`) can never have default values;
+    /// non-variadic parameters sometimes will.
+    pub fn iter_non_variadic_params(&self) -> impl Iterator<Item = &ParameterWithDefault> {
         self.posonlyargs
             .iter()
             .chain(&self.args)
             .chain(&self.kwonlyargs)
+    }
+
+    /// Returns the [`ParameterWithDefault`] with the given name, or `None` if no such [`ParameterWithDefault`] exists.
+    pub fn find(&self, name: &str) -> Option<&ParameterWithDefault> {
+        self.iter_non_variadic_params()
             .find(|arg| arg.parameter.name.as_str() == name)
     }
 
-    /// Returns `true` if a parameter with the given name included in this [`Parameters`].
+    /// Returns an iterator over all parameters included in this [`Parameters`] node.
+    pub fn iter(&self) -> ParametersIterator {
+        ParametersIterator::new(self)
+    }
+
+    /// Returns the total number of parameters included in this [`Parameters`] node.
+    pub fn len(&self) -> usize {
+        self.posonlyargs.len()
+            + self.args.len()
+            + usize::from(self.vararg.is_some())
+            + self.kwonlyargs.len()
+            + usize::from(self.kwarg.is_some())
+    }
```

Comment:
Oh, I suppose this could theoretically overflow, given that Python functions can have an arbitrary number of parameters, so the number of parameters could theoretically exceed `usize::MAX`. This isn't a new issue -- we were already doing this calculation in `crates/ruff_python_formatter/src/other/parameters.rs` and `crates/ruff_python_parser/src/parser/statement.rs`. But maybe we should do `checked_add()` here and return an `Option`?

The formatter function that uses this method returns a `Result`, so we should be able to gracefully handle an `Option` there. The parser function that calls this might be slightly trickier to refactor.

Reply:
I would need to look into the supported platforms. But I think it's safe to say that it's impossible that this overflows for files that are <= 4GB and the platform uses at least a `u32` as `usize`.